src/Grid/topology.jl

```diff
@@ -0,0 +1,341 @@
+the given entity is included in the returned list as well.
+"""
+getneighborhood
+
+struct EntityNeighborhood{T<:Union{BoundaryIndex,CellIndex}}
+    neighbor_info::Vector{T}
+end
+
+EntityNeighborhood(info::T) where T <: BoundaryIndex = EntityNeighborhood([info])
+Base.zero(::Type{EntityNeighborhood{T}}) where T = EntityNeighborhood(T[])
+Base.zero(::Type{EntityNeighborhood}) = EntityNeighborhood(BoundaryIndex[])
+Base.length(n::EntityNeighborhood) = length(n.neighbor_info)
+Base.getindex(n::EntityNeighborhood,i) = getindex(n.neighbor_info,i)
+Base.firstindex(n::EntityNeighborhood) = 1
+Base.lastindex(n::EntityNeighborhood) = length(n.neighbor_info)
+Base.:(==)(n1::EntityNeighborhood, n2::EntityNeighborhood) = n1.neighbor_info == n2.neighbor_info
+Base.iterate(n::EntityNeighborhood, state=1) = iterate(n.neighbor_info,state)
+
+function Base.:+(n1::EntityNeighborhood{T1}, n2::EntityNeighborhood{T2}) where {T1 <:Union{BoundaryIndex,CellIndex}, T2 <:Union{BoundaryIndex,CellIndex}}
+    neighbor_info = [n1.neighbor_info; n2.neighbor_info]
+    return EntityNeighborhood(neighbor_info)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", n::EntityNeighborhood)
+    if length(n) == 0
+        println(io, "No EntityNeighborhood")
+    elseif length(n) == 1
+        println(io, "$(n.neighbor_info[1])")
+    else
+        println(io, "$(n.neighbor_info...)")
+    end
+end
+
+abstract type AbstractTopology end
+
+"""
+    ExclusiveTopology(cells::Vector{C}) where C <: AbstractCell
+`ExclusiveTopology` saves topological (connectivity/neighborhood) data of the grid. The constructor works with an `AbstractCell`
+vector for all cells that dispatch `vertices`, `faces` and in 3D `edges`.
+The struct saves the highest dimensional neighborhood, i.e. if something is connected by a face and an
+ edge only the face neighborhood is saved. The lower dimensional neighborhood is recomputed, if needed.
+
+# Fields
+- `vertex_to_cell::Dict{Int,Vector{Int}}`: global vertex id to all cells containing the vertex
+- `cell_neighbor::Vector{EntityNeighborhood{CellIndex}}`: cellid to all connected cells
+- `face_neighbor::SparseMatrixCSC{EntityNeighborhood,Int}`: `face_neighbor[cellid,local_face_id]` -> neighboring face
+- `vertex_neighbor::SparseMatrixCSC{EntityNeighborhood,Int}`: `vertex_neighbor[cellid,local_vertex_id]` -> neighboring vertex
+- `edge_neighbor::SparseMatrixCSC{EntityNeighborhood,Int}`: `edge_neighbor[cellid_local_vertex_id]` -> neighboring edge
+"""
+struct ExclusiveTopology <: AbstractTopology
+    # maps a global vertex id to all cells containing the vertex
+    vertex_to_cell::Vector{Set{Int}}
+    # index of the vector = cell id ->  all other connected cells
+    cell_neighbor::Vector{EntityNeighborhood{CellIndex}}
+    # face_neighbor[cellid,local_face_id] -> exclusive connected entities (not restricted to one entity)
+    face_face_neighbor::Matrix{EntityNeighborhood{FaceIndex}}
```

Comment:
Why a dense matrix over a sparse one?

Reply:
It seems to be a bit faster and uses less memory, because we had almost all entries filled in the sparse matrix (which is above the rule-of-thumb threshold of nlog(n) entries).

Reply:
> because we had almost all entries filled in the sparse matrix

ah I understand the misconception I had. I thought with increasing cells, there should be less non-connectivities encoded. However, non-connectivies are encoded by not having more entries in the EntityNeighborhood struct and certainly each cell does have a connectivity except for the boundary elements. So this makes absolute sense